### PR TITLE
Testspec

### PR DESCRIPTION
**Pull request by jedwards4b**
With this change we provide an initial version of the testspec.xml file used by cesm and we allow the user to optionally provide arguments to the --generate and --compare options.   generate and compare can now be used together if and only if they point to different directories.

### CODE REVIEW

File: utils/python/CIME/XML/generic_xml.py [hunk 1]
```diff
@@ -51,11 +51,12 @@ def write(self, infile=None):
         """
         Write an xml file from data in self
         """
+        if(infile is None):
+            infile = self.filename
         logging.info("write: "+ infile)
-        if(infile != None):
```
Comment (jgfouca) [inline utils/python/CIME/XML/generic_xml.py]: What was the motivation for this change? Outputted XML didn't look good?

File: utils/python/CIME/create_test.py [hunk 2]
```diff
@@ -95,17 +98,33 @@ def __init__(self, test_names,
         else:
             self._parallel_jobs = parallel_jobs
 
-        if (self._compare or self._generate):
+        self._baseline_cmp_name = None
+        self._baseline_gen_name = None
 
+        if (self._compare or self._generate):
+            logging.warn("compare is %s generate is %s"%(self._compare,self._generate))
```
Comment (jgfouca) [inline utils/python/CIME/create_test.py]: Probably should be logging.info instead of logging.warn

File: utils/python/CIME/create_test.py [hunk 3]
```diff
@@ -95,17 +98,33 @@ def __init__(self, test_names,
...
             # Figure out what baseline name to use
             if (baseline_name is None):
+                if(self._compare is not None and self._compare == str(self._compare)):
```
Comment (jgfouca) [inline utils/python/CIME/create_test.py]: _compare and _generate should always be booleans. If the incoming compare argument is a string, set self._compare to True and self._baseline_cmp_name to compare argument.

File: utils/python/CIME/create_test.py [hunk 4]
```diff
@@ -228,10 +251,12 @@ def _update_test_status(self, test, phase, status):
                    "New phase should be set to pending status")
             expect(self._phases.index(old_phase) == phase_idx - 1,
                    "Skipped phase?")
-
+        if(self._cime_model == "cesm"):
```
Comment (jgfouca) [inline utils/python/CIME/create_test.py]: Does this really need to be updated every time a state changes or can it just be updated at the end?

File: utils/python/wait_for_tests.py [hunk 5]
```diff
@@ -429,6 +429,12 @@ def wait_for_tests_impl(test_paths, no_wait=False, check_throughput=False, check
 
     return test_results
 
+def update_testspec_xml(test_results):
```
Comment (jgfouca) [inline utils/python/wait_for_tests.py]: I'm confused by this on many levels. Wait_for_tests already prints statuses. Also, why is this function called "update_testspec_xml" when it's just printing testnames?

File: utils/python/tests/scripts_regression_tests [hunk 6]
```diff
@@ -416,7 +417,8 @@ class TestCreateTestImpl(TestCreateTestCommon):
         for idx, phase in enumerate(ct._phases):
             for test in ct._tests:
                 if (phase == CIME.create_test.INITIAL_PHASE):
-                    continue
+                    if (ct._cime_model == "cesm"):
```
Comment (jgfouca) [inline utils/python/tests/scripts_regression_tests]: Testing phases shouldn't have anything to do with testspec.

File: utils/python/CIME/XML/generic_xml.py (hunk 1 above)
Comment (jedwards4b) [inline utils/python/CIME/XML/generic_xml.py]: Yes, when I tried to output the internally generated testspec.xml file it was alll on a single line with no formatting.  I would really like to use the xslt library here but the python interface doesn't seem to be generally available.

File: utils/python/CIME/create_test.py (hunk 2 above)
Comment (jedwards4b) [inline utils/python/CIME/create_test.py]: fixed

File: utils/python/CIME/create_test.py (hunk 3 above)
Comment (jedwards4b) [inline utils/python/CIME/create_test.py]: fixed

File: utils/python/wait_for_tests.py (hunk 5 above)
Comment (jedwards4b) [inline utils/python/wait_for_tests.py]: this is incomplete code - i guess I got distracted and forgot about it.  Maybe I'll pull the testspec changes out of this PR and do it separately.

**Comment by jgfouca** [comment]
Review complete.

File: utils/python/CIME/XML/generic_xml.py (hunk 1 above)
Comment (jgfouca) [inline utils/python/CIME/XML/generic_xml.py]: That's odd. When I use generic_xml to write XML files, they seem to look OK.

File: utils/python/CIME/create_test.py [hunk 7]
```diff
@@ -95,17 +95,37 @@ def __init__(self, test_names,
...
-
+        self._baseline_cmp_name = None
+        self._baseline_gen_name = None
+        self._compare = False
+        self._generate = False
+        if (compare or generate):
+            logging.info("compare is %s generate is %s"%(compare,generate))
             # Figure out what baseline name to use
             if (baseline_name is None):
+                if(compare is not None and compare == str(compare)):
```
Comment (jgfouca) [inline utils/python/CIME/create_test.py]: To check if compare is a string, you want this: "type(compare) is str"

File: utils/python/wait_for_tests.py (hunk 5 above)
Comment (jgfouca) [inline utils/python/wait_for_tests.py]: Did you intend to remove this in your recent remove test-spec commit?

**Comment by jgfouca** [comment]
Getting close. Only string type checking in create_test and testspec in wait_for_tests need to be fixed.

File: utils/python/CIME/create_test.py (hunk 7 above)
Comment (bandre-ucar) [inline utils/python/CIME/create_test.py]: You should use the isinstance function: if isinstance(compare, str)

Comment (jgfouca) [inline utils/python/CIME/create_test.py]: Yeah, pylint flags type(X) is Y and tells me to use isinstance instead. I kinda disagree for primitive types like str though, the former just reads better to me.

**Pull request merged**